File: java/56-merge-intervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> ans = new ArrayList<>();
        if (intervals == null || intervals.size() == 0) {
            return ans;
        }
        Collections.sort(intervals, (i1, i2) -> {
            return Integer.compare(i1.start, i2.start);
        });
        Interval last = null;
        for (Interval curr : intervals) {
            if (last == null) {
                last = curr;
            } else {
                if (last.end >= curr.start) {
                    // merge
                    last.end = Math.max(last.end, curr.end);
                } else {
                    ans.add(last);
                    last = curr;
                }
            }
        }
        ans.add(last);
        return ans;
    }
}

File: python/56-merge-intervals.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def merge(self, intervals: List[Interval]) -> List[Interval]:
        if intervals is None or len(intervals) <= 1:
            return intervals
        intervals = sorted(intervals, key=lambda interval: interval.start)
        prev = intervals[0]
        ans = []
        for i in range(1, len(intervals)):
            curr = intervals[i]
            if prev.end < curr.start:
                ans.append(prev)
                prev = curr
            else:
                prev.start = min(prev.start, curr.start)
                prev.end = max(prev.end, curr.end)
        ans.append(prev)
        return ans